### ISSUE TITLE
Game Data Storage

### ISSUE THREAD

**Issue by minz1** [ugdt/gdxcrawler#10]
We must have some sort of storage system for things such as tile properties, or the entire world.
We must find a library to facilitate database design in C#, and implement storage for the game world and tile properties.

**Comment by minz1** [comment]
I'm thinking we use Dapper for its speed, since we want to be able to live load a chunk.

**Comment by minz1** [comment]
Regarding DB structure,
Each game save is stored in its own database. This database will have a World table, which has a 1 to many relationship to the Chunk table. The Chunk table will have a 1 to many relationship with a Tile table.

The World table will store the chunks. Somewhere, we should store the seed, but I'm not sure on where exactly to do that. Perhaps an extra column, but that is unnecessary space. We could have a Game table that holds the world perhaps, and the world is stored with a seed column.

The Chunk table will store the Tiles, and maybe entities on the tile? I'm not sure on how this will interconnect with ugdt/zelcrawler#6.

Tiles will store their tileId, and any properties for the tile will be saved as a json string, maybe. I don't know the specifics of how that will happen, this depends on ugdt/zelcrawler#5.

**Comment by abueide** [comment]
I agree on the choice to use dapper if we go with sqlite. Entity Framework seems a bit higher up and would probably affect the architecture of our database significantly.

I don't like storing properties into a json string in the database, but I'm not directly opposed to it. For game development I think we're going to run into a lot of similar issues down the line, a NOSQL database such as mongodb or redis might actually be more suited to our needs. We're going to have a lot of loosely structured data that might not fit very well into a relational database. Having an easy way to just convert our item, tile & game object definitions then store them in a nosql database seems like it could be the way to go.

See: 
https://stackoverflow.com/questions/32325390/how-to-store-a-dynamic-python-dictionary-in-mysql-database

https://gamedev.stackexchange.com/questions/5316/nosql-is-it-a-valid-option-for-web-based-game

**Comment by abueide** [comment]
The problem is we can't just use the popular mongodb & redis, we need something akin to sqlite on the nosql side. We're looking for whats called an "embeddable nosql database".

We need to do more research, but litedb in particular sticks out
https://www.litedb.org/

**Comment by minz1** [comment]
I think we could make it work with a relational database. The thing is, we do have a lot of structured data, with the relationships between tiles and chunks, worlds and levels, it makes more sense to use sql than nosql imo.

**Comment by minz1** [comment]
https://medium.com/dapper-net/custom-type-handling-4b447b97c620

**Comment by minz1** [comment]
The previous conversation on this topic is no longer as relevant. We now have options for game state saving, by persisting the ECS and TileMap.

**Comment by abueide** [comment]
@minz1 i'm still thinking exposed & sqlite for things like item & tile definitions since the preferences api doesn't really make sense.

TileWorld should be injected into the game world that way systems have access to interact with the world. TileWorld should become the api for interacting with the environment. As a bonus when we serialize the artemis world, the tileworld should get serialized with it (if the serialization thing handles injected stuff too).

So here are some considerations for how we should persist different types of data:
Game state: Serialize into binary format using artemis
Game Settings: Should have different classes for different things (e.g. InputSettings, RenderSettings) that gets stored under a Profile class. Each Profile will be stored using libgdx's preferences api.
Game data (entity definitions, tile definitions, item definitions etc): preferences api doesn't make sense since it doesn't change on the fly. It shouldn't be saved with the Game State for the same reason. We have a few options here: json, xml, sqlite, or plain kotlin classes. No matter which format we choose, we'll probably have ingame or external tools to generate/modify the game data so we won't have to write json, xml or modify sqlite db by hand.  We could also make those tools generate/modify enums if we went the kotlin route. I'm imagining for entities we'll have a tool that can access all the components we wrote and you'll be able to select components that will get instantiated through reflection when the entity archetype is loaded. And then the archetypes would be stored in the external format we choose.

I'm leaning towards sqlite, because it will make it much easier when I switch over to multiplayer later on, as well as like hot reloading assets without recompiling. Can probably be done with files if we don't put them in the assets folder and instead put them in the filesystem, but seems like more of a pain.
